[MLIR][SCFToGPU] Guard operands before AffineApplyOp::create to avoid crash

shashforge opened the pull request:
This fixes a crash in SCF→GPU when building the per‑dim index for mapped scf.parallel.

**Change**:
- Map step/lb through cloningMap, then run ensureLaunchIndependent.
- If either is still unavailable at launch scope, emit a match‑failure; otherwise build the affine.apply.

**Why this is correct:**
- Matches how the pass already handles launch bounds; avoids creating an op with invalid operands and replaces a segfault with a clear diagnostic.

**Tests**:
- Added two small regressions that lower to gpu.launch and exercise the affine.apply path.

Fixes :  #167654

shashforge committed:
[MLIR][SCFToGPU] Guard operands before AffineApplyOp::create to avoid crash

When lowering mapped scf.parallel to gpu.launch, the per-dimension index
was built with AffineApplyOp::create using ensureLaunchIndependent(lb/step)
directly. If lb/step were not available at the launch scope, that helper
returned an empty value and the builder crashed while creating the op.

Mirror the bound-handling path: first map lb/step through cloningMap, then
call ensureLaunchIndependent. If either operand still isn’t available above
the launch, report a precise match-failure instead of crashing. This makes
conversion fail cleanly on invalid cases and succeed for valid ones.

Add two positive regressions that previously crashed and now lower to
gpu.launch (and affine.apply).

Fixes: #167654
Signed-off-by: Shashi Shankar <shashishankar1687@gmail.com>

shashforge committed:
Merge branch 'main' into mlir-scf2gpu-affineapply-crash

shashforge committed:
Merge branch 'main' into mlir-scf2gpu-affineapply-crash

linuxlonelyeagle commented on mlir/test/Conversion/SCFToGPU/parallel-to-gpu-crash-regression.mlir:
```diff
@@ -0,0 +1,31 @@
+// and ensure we see a gpu.launch and an affine.apply (no crash).
+
+module {
+  func.func @two_dim_parallel_mapped() {
```
Please add the test in the `mlir/test/Conversion/SCFToGPU`

shashforge replied:
@linuxlonelyeagle  Done. I moved the tests into mlir/test/Conversion/SCFToGPU/parallel_loop.mlir

linuxlonelyeagle reviewed (changes requested):
Some test-related suggestions.

shashforge committed:
Merge branch 'main' into mlir-scf2gpu-affineapply-crash

Signed-off-by: Shashi Shankar <shashishankar1687@gmail.com>

shashforge commented:
@linuxlonelyeagle @joker-eph  — thanks for the earlier pointers.
I moved the tests into `mlir/test/Conversion/SCFToGPU/parallel_loop.mlir` and simplified the checks per the Testing Guide.
If you have a minute to re‑review, I’d appreciate it. I’m happy to make further tweaks.

shashforge committed:
Merge branch 'main' into mlir-scf2gpu-affineapply-crash

shashforge committed:
Merge branch 'main' into mlir-scf2gpu-affineapply-crash

Signed-off-by: Shashi Shankar <shashishankar1687@gmail.com>

linuxlonelyeagle commented on mlir/test/Conversion/SCFToGPU/parallel_loop.mlir:
```diff
@@ -673,3 +673,51 @@ func.func @nested_parallel_with_side_effect() {
+// CHECK-LABEL: func.func @scf2gpu_index_creation_2d
+// CHECK:       gpu.launch
+// CHECK:       affine.apply
+
```
This is a good example https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Vector/canonicalize/vector-from-elements.mlir.

linuxlonelyeagle reviewed (changes requested):
Some changes required to the formatting.

shashforge committed:
Merge branch 'main' into mlir-scf2gpu-affineapply-crash

joker-eph reviewed (approved):
LGTM, but please wait for @linuxlonelyeagle to have another look

shashforge committed:
Merge branch 'main' into mlir-scf2gpu-affineapply-crash

Signed-off-by: Shashi Shankar <shashishankar1687@gmail.com>

linuxlonelyeagle reviewed (commented):
A few more details to note:
* You can align the colon.
* Use spaces to ensure the hierarchy of IR.
```
// CHECK-LABEL: func @scf2gpu_index_creation_2d
//       CHECK:   gpu.launch
//       CHECK:     %[[IDX:.*]] = affine.apply
//       CHECK:     arith.addi %[[IDX]],
```

shashforge committed:
Merge branch 'main' into mlir-scf2gpu-affineapply-crash

Signed-off-by: Shashi Shankar <shashishankar1687@gmail.com>

linuxlonelyeagle reviewed (approved):
LGTM, Thank you for your efforts. Please tell me, would you like me to help you merge the code?
